control-plane/api-gateway/binding/cleanup.go

```diff
@@ -0,0 +1,198 @@
+package binding
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	mapset "github.com/deckarep/golang-set/v2"
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/consul-k8s/control-plane/consul"
+	"github.com/hashicorp/consul/api"
+)
+
+const (
+	oldACLRoleName   = "managed-gateway-acl-role"
+	oldACLPolicyName = "api-gateway-token-policy"
+)
+
+var sleepTime = 60 * time.Second
+
+type Cleaner struct {
+	Logger       logr.Logger
+	ConsulConfig *consul.Config
+	ServerMgr    consul.ServerConnectionManager
+	AuthMethod   string
+}
+
+// Run periodically cleans up old ACL roles and policies as well as orphaned inline certs
+// when it detects that there are no more inline certs and the old ACL role and policy are not in use then it exits.
+func (c Cleaner) Run(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(sleepTime):
+		}
+
+		client, err := consul.NewClientFromConnMgr(c.ConsulConfig, c.ServerMgr)
+		if err != nil {
+			c.Logger.Error(err, "failed to create Consul client")
+			continue
+		}
+
+		aclsCleanedUp, err := c.cleanupACLRoleAndPolicy(client)
+		if err != nil {
+			c.Logger.Error(err, "failed to cleanup old ACL role and policy")
+		}
+
+		inlineCertsAllCleanedUp, err := cleanupInlineCerts(client)
+		if err != nil {
+			c.Logger.Error(err, "failed to cleanup inlineCerts")
+		}
+
+		if aclsCleanedUp && inlineCertsAllCleanedUp {
+			return
+		}
+	}
+}
+
+// cleanupACLRoleAndPolicy deletes the old shared gateway ACL role and policy if they exist.
+func (c Cleaner) cleanupACLRoleAndPolicy(client *api.Client) (bool, error) {
+	existingRules, _, err := client.ACL().BindingRuleList(c.AuthMethod, &api.QueryOptions{})
```

Comment:
Does this list for all namespaces?

Reply:
Discussed on Slack that the gateway approach described below is probably more scalable in all use cases and would query directly whatever namespace the gateway resided in.

Reply:
we also always write the binding-rules into the default namespace and partition https://github.com/hashicorp/consul-k8s/blob/main/control-plane/api-gateway/cache/consul.go#L557